Add biscuit module

zwdzwd opened the pull request:
This adds support for QC output from BISCUIT. This includes: QC for mapping uniformity, retention rates, duplication uniformity, CpG coverage, methylation bias, mapping summary, etc. Initial testing should pass with the test data on MultiQC_TestData (also a pull request there). Please review. Thank you.

zwdzwd commented:
Hi @ewels, Thanks for the feedbacks. They are very helpful. I will go work on those and get back to you.

ewels commented:
Hi @zwdzwd,

Have you had a chance to make any progress on this PR recently?

Phil

zwdzwd commented:
@ewels 
Sorry for the late response. I got swamped by other things. I am still working on this. I will try get back to you by next week.

zwdzwd commented:
Hi @ewels ,
My apology again for delayed action. Could you please check this version? I managed to cut the number of sections to 11 by combining figures to tabs and combining tables. I removed a couple of sections of less importance. I wonder if this may be an OK number for initial implementation? I also double checked to make sure titles and axis labels are there and sample names are clean. Let me know if I am missing something or there is a better way of doing things. 

Thank you for making such a handy and beautiful tool!

Wanding

ewels commented:
Great stuff, thanks! No problem with the delay, I’ve been very busy recently too. I will spend a day working on MultiQC later this week and have another look then.

Thanks!

Phil

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+
+        # Find and parse bismark alignment reports
+        for k in self.mdata:
+            for f in self.find_log_files('biscuit/%s' % k):
```
Minor thing, but I'd prefer format strings if possible - this is the code style that's used everywhere else in MultiQC. eg:

```python
'biscuit/{}'.format(k)
```

This doesn't make any functional difference, but helps with consistency :)

zwdzwd replied:
I changed the code to this style.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+        num_parsed = 0
+        for k in self.mdata:
+            num_parsed += len(self.mdata[k])
+        if num_parsed == 0:
```
A slightly more condensed version could be:

```python
if sum([len(self.mdata[k]) for k in self.mdata]) == 0:
```

zwdzwd replied:
Thanks for showing me. This version is used.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+        for k in self.mdata:
+            if len(self.mdata[k]) > 0:
+                self.write_data_file(self.mdata[k], 'multiqc_biscuit_%s' % k)
+                log.info("Found %d biscuit %s reports" % (len(self.mdata[k]), k))
```
The log message is prefixed with the module name (biscuit), so no need to repeat it here.

zwdzwd replied:
The repetition is removed.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+                self.write_data_file(self.mdata[k], 'multiqc_biscuit_%s' % k)
+                log.info("Found %d biscuit %s reports" % (len(self.mdata[k]), k))
+                if k != 'markdup':
+                    getattr(self, 'chart_biscuit_%s' % k)()
```
What does this do?

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+        for sid, dd in self.mdata['mapq'].items():
+            if not sid.endswith('_mapq_table.txt'):
+                continue
+            sid = sid.replace('_mapq_table.txt','')
```
Please run through the `self.clean_s_name()` function as well. This is needed for the `-b` function and custom sample name cleanups.

See http://multiqc.info/docs/#custom-sample-names

ewels replied:
Ah, I just got to the search pattern config. Yes, so it's better to have a single search pattern config for every _type_ of file rather than relying on the filename for this reason. Then if each file type is parsed into a separate data structure, you shouldn't need to have this check.

zwdzwd replied:
I added clean_s_name
```
s_name = self.clean_s_name(f['fn'], f['root']) # allow user to specify directory
self.mdata[k][s_name] = d
```

I did my best to separate search pattern for each file type and removed existing file checking best as I can.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+
+        pd = {}
+        for sid, dd in self.mdata['mapq'].items():
+            if not sid.endswith('_mapq_table.txt'):
```
I'd prefer to remove this. Sometimes MultiQC users change the filenames of tool outputs and then customise their MultiQC config to update the search patterns. This will break things for them as the filename could not have that string in for genuine reasons.

If you want to validate that the correct thing has been parsed, I'd recommend doing this in the parsing step instead (preferably by checking file contents / parsed data).

zwdzwd replied:
These are removed as mentioned above.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            pheader['%duprateSE'] = {'title':'% DupsSE', 'max':100, 'min':0, 'suffix':'%','scale':'Reds'}
+        self.general_stats_addcols(pd, pheader)
+
+        # if len(mdata)>0 and all([dd['Duplication Rate'] is not None for dd in mdata.values()]):
```
Please remove old commented out code.

zwdzwd replied:
The old commented code has been removed.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+    def parse_logs_mapq(self, f, fn):
+
+        data = {}
+        if fn.endswith('_mapq_table.txt'):
```
Yup, please split each of these out into their own function with their own search pattern.

zwdzwd replied:
Done.:)

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            name = 'Mapping Summary',
+            anchor = 'biscuit-mapping',
+            plot = bargraph.plot(pd, OrderedDict([
+                ('OAligned', {'name':'Optimally Aligned Reads', 'color': '#8bbc21'}),
```
Personally, I prefer charts to use colours from the existing MultiQC colour palette (the defaults here should actually be pretty good I think? Blue / Dark Blue / Orange off the top of my head). But it's your code, so if you really prefer these colours then it's fine to stick with them 😉

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+
+        self.add_section(
+            name = 'Mapping Summary',
+            anchor = 'biscuit-mapping',
```
Please add at least a short `description` and preferably a longer `helptext` to describe what to expect and what to look out for.

zwdzwd replied:
The description and helptext have been added.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            ]), {'id':'biscuit_strands',
+                 'title':'BISCUIT: Mapping Strand Distribution',
+                 'ylab':'Number of Reads',
+                 'cpswitch_c_active': True,
```
`True` is the default - maybe you mean `False` so that the bars display percentages by default? This would make sense to me, as we're looking for differences in proportions here, rather than counts.

zwdzwd replied:
Yes, I changed it to False. Thanks

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            anchor='biscuit-strands',
+            description = "<p>This plot shows the distribution of strand of mapping and strand of bisulfite conversion.</p>",
+            plot = bargraph.plot(pd, OrderedDict([
+                ('++', {'name':'Waston-Aligned, Waston-Bisulfite Conversion', 'color': '#F53855'}),
```
Same comment with colours as above, but fine to leave if you prefer.

zwdzwd replied:
I leave the color of this one as it appears to highlight the contrast better. I think the color code I was using comes from the Documentation IIRC.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+        self.add_section(
+            name='Mapping Strand Distribution',
+            anchor='biscuit-strands',
+            description = "<p>This plot shows the distribution of strand of mapping and strand of bisulfite conversion.</p>",
```
The description is parsed as markdown, so no need for HTML tags such as `<p>`.

It would be great if you could write a `helptext` talking about different possibilities here and what to expect. eg. different types of bisulfite library prep.

zwdzwd replied:
"<p>" are removed. helptext is added.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            anchor = 'biscuit-mapq-isize',
+            description = "<p>This plot shows the distribution of Primary Mapping Quality.</p>",
+            plot = linegraph.plot([pd_mapping, pd_isize],
+                {'id':'biscuit_mapping', 'title': 'Mapping Information', 'data_labels': [
```
Please prepend the title with the module name.

ewels replied:
Suggested additional config options:
```python
'ymin': 0,
'yCeiling': 100,
'yMinRange': 10,
'yLabelFormat': '{value}%',
'tt_label': '<strong>Q{point.x}:</strong> {point.y:.2f}% of reads'
```

zwdzwd replied:
These are added accordingly.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            plot = linegraph.plot([pd_mapping, pd_isize],
+                {'id':'biscuit_mapping', 'title': 'Mapping Information', 'data_labels': [
+                    {'name':'Mapping Quality', 'ylab': '% Primary Mapped Reads','xlab': 'Mapping Quality'},
+                    {'name':'Insert Size', 'ylab': '% Mapped Reads', 'xlab': 'Insert Size'}]}))
```
This graph is empty for me when I test. Please can you add code to check that there is data to plot so that the tab isn't added if there is no data..

ewels replied:
Also, I know I asked you to combine plots and reduce number of sections, but I don't really see how insert size and mapping quality are related. So I think these _should_ be separate sections and plots. Sorry!

zwdzwd replied:
I think the graph is empty because of the obsolete test data. I separated insert size and mapping quality section. Thank you.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            self.add_section(
+                name = 'Cumulative Base Coverage',
+                anchor = 'biscuit-coverage-base',
+                description = "<p>This plot shows the cummulative base coverage. High and low GC content region are the top and bottom 10% 100bp window in GC content.</p>",
```
No HTML tags needed. Please also describe what the Q means. Help text would also be, well, helpful!

zwdzwd replied:
`p` is removed. I added help text.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+                anchor = 'biscuit-coverage-base',
+                description = "<p>This plot shows the cummulative base coverage. High and low GC content region are the top and bottom 10% 100bp window in GC content.</p>",
+                plot = linegraph.plot(mdata, {'id':'biscuit_coverage_base',
+                    'title': 'Cumulative Base Coverage',
```
Please add an X axis, `xLabelFormat` with the `X` suffix to denote coverage,  prepend the title with the module name.

zwdzwd replied:
X-axis and xLabelFormat are added as suggested.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+                plot = linegraph.plot(mdata, {'id':'biscuit_coverage_base',
+                    'title': 'Cumulative Base Coverage',
+                    'data_labels': [
+                        {'name': 'All', 'ylab':'Million Bases'}, 
```
Please adjust the ylab values to reflect which plot is shown. This is helpful when single plots are exported in isolation.

zwdzwd replied:
Now ylab values are made unique to reflect the current plot shown. Thanks for the good idea.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            for sid, dd in mdata[0].items():
+                if sid not in basecov:
+                    basecov[sid] = {}
+                basecov[sid]['all'] = dd[1]/dd[0]*100
```
Note that you want to be careful with ints and floats here. `100` is an `int`, so if the others are parsed as ints then the result will also be an integer. I usually wrap stuff in `float()` and use `100.0` to force floating number results.

zwdzwd replied:
Thanks for the warning. I added `float()` and `100.0` to force floating number calculation.

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+            self.add_section(
+                name = 'Coverage by At Least One Read',
+                anchor = 'biscuit-coverage-base-table',
+                description = '<p>The fraction covered by at least one read.</p>',
```
The fraction of what? No HTML tags needed.

zwdzwd replied:
`<p>` removed. The description is updated to clarify it's the fraction of either genome or genomic CpGs (depends on tab).

ewels commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,688 @@
+        self.add_section(
+            name = 'Cytosine Retention',
+            anchor = 'biscuit-retention',
+            description = "<p>This plot shows cytosine retention rate. 'r.' stands for read-averaging and 'b.' stands for 'base-averaging.</p>",
```
You can use markdown backticks for `r` which will look nicer that 'r' I think. No HTML tags needed.

ewels replied:
It would be great to have help text explaining this too. I've worked with bisulfite data quite a bit and I'm not sure what cytosine retention means.

zwdzwd replied:
Thanks for the suggestion. I switched to `r` and removed `<p>`. I added help text to explain on cytosine retention/conversion.

ewels commented on multiqc/utils/search_patterns.yaml:
```diff
@@ -82,6 +82,32 @@ biobloomtools:
     num_lines: 2
 bismark/align:
     fn: '*_[SP]E_report.txt'
+biscuit/mapq:
```
File search patterns should be unique for each _type_ of data. The way that the code is written currently makes it impossible for people to customise these search patterns if they modify their data files, as the module code expects certain file paths. So instead, each file type should have its own search pattern:

```yaml
biscuit/mapq_mapq_table: fn: '*_mapq_table.txt'
biscuit/mapq_strand_table: fn: '*_strand_table.txt'
biscuit/mapq_isize_score_table: fn: '*_isize_score_table.txt'
```

zwdzwd replied:
Yes, now each file type has its own search pattern.

ewels commented on multiqc/utils/search_patterns.yaml:
```diff
@@ -82,6 +82,32 @@ biobloomtools:
+    - fn: '*_strand_table.txt'
+    - fn: '*_isize_score_table.txt'
+biscuit/markdup:
+    fn: '*_markdup_report.txt'
```
This is quite a generic filename which could conceivably also be used by other tools. It would be good to also add a `content` string to confirm the identity of the file. Otherwise, even if the module code ignores it, the file will not be used by any other modules.

zwdzwd replied:
I updated the log file format to include a (quite) distinguishable header (BISCUITqc). I updated both search_pattern.yml and the script to detect this header. I hope this reduces the cross-talk between modules. Thanks for pointing me to `content`.

ewels commented on multiqc/utils/search_patterns.yaml:
```diff
@@ -82,6 +82,32 @@ biobloomtools:
+    - fn: '*_covdist_table.txt'
+    - fn: '*_cpg_cv_table.txt'
+    - fn: '*_cpg_dist_table.txt'
+    - fn: '*_dup_report.txt'
```
Same as above with the common filename pattern.

zwdzwd replied:
Same solution as in response above. Thanks.

ewels commented:
Two more things: I only added one comment about using `self.clean_s_name()`, but that should be used every time a sample name is generated (this is quite a few places in the code). Also, parsed data structures should be filtered using `self.ignore_samples()` so that users can choose to remove certain samples.

zwdzwd commented on .travis.yml:
```diff
@@ -10,7 +10,7 @@ matrix:
 install:
   - python setup.py install
-  - git clone https://github.com/ewels/MultiQC_TestData.git test_data
+  - git clone https://github.com/zwdzwd/MultiQC_TestData.git test_data
```
This is reverted.

zwdzwd commented:
Thank you @ewels for making such an extensive review. I have addressed most of the comments which actually led to quite some refactoring of both the multiqc script and biscuit QC script itself. Yes, my PI did like the multiqc very very much and intended to use it as the primary reporting tool for Biscuit. The biggest change in the current code is to have each file type identified by a single search pattern. This indeed removed some previous check inside the script code to distinguish file type. I also added a few descriptions and help text as suggested. I will probably add more as the review goes on. I also added quite some annotation inside the script to make it more readable than the previous version.

Thanks again for your instruction, suggestions and feedback. Please certainly let me know of anything that I may still miss.

Wanding

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+import re
+
+from multiqc import config
+from multiqc.plots import beeswarm, linegraph, bargraph, table
```
Also removed. Thanks.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+            'retention_rate_bybase': {},
+        }
+
+        # Find and parse bismark alignment reports
```
Sorry. My oversight. Now removed. :)

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+        for k in self.mdata:
+            for f in self.find_log_files('biscuit/{}'.format(k)):
+                d = getattr(self, 'parse_logs_%s' % k)(f['f'], f['fn'])
+                s_name = self.clean_s_name(f['fn'], f['root']) # allow user to specify directory
```
This make sense, I still kept the `s_name = self.clean_s_name` part for a later comment. This is the place sample names were cleaned first round. They get more cleaning (in ad hoc ways) in each functions.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+        hasPE = False
+        hasSE = False
+        for sid, dd in self.mdata['markdup_report'].items():
+            sid = sid.replace('_markdup_report','')
```
That makes sense. As mentioned above, I now have "clean_s_name" at the time of parsing. The plotting code does contain some more cleaning but they tend to be targeting uncommon suffixes.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+        pheader = OrderedDict()
+        pheader['%aligned'] = {'title':'% Aligned', 'max':100, 'min':0, 'suffix':'%','scale':'Greens'}
+        if hasPE:
+            pheader['%dupratePE'] = {'title':'% DupsPE', 'max':100, 'min':0, 'suffix':'%','scale':'Reds'}
```
Good information. It is deleted now.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+            ]), {'id':'biscuit_mapping_summary',
+                 'title':'BISCUIT: Mapping Summary',
+                 'ylab':'Number of Reads',
+                 'cpswitch_c_active': False,
```
Sure, I deleted this for consistency.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+        # Mapping Quality together in one plot
+        total = {}
+        for sid, dd in self.mdata['align_mapq'].items():
+            sid = sid.replace('_mapq_table','')
```
That makes sense. I regrouped the name cleaning operations at parsing.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+            plot = linegraph.plot(pd, {
+                'id': 'biscuit_retention_read_cpa', 
+                'xlab': 'Number of Retention within Read',
+                'title': 'Retention Distribution',
```
Added. Thanks.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+            plot = linegraph.plot(mdata, {
+                'id': 'biscuit_retention_cytosine',
+                'xlab': 'Position in Read', 'ymin':0, 'ymax':100, 'yMinRange':0, 'yFloor':0,
+                'title': 'Retention vs. Base Position in Read',
```
Added. Thanks.

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+        pheader['cv_cpg_botgc'] = {'title':'CG.CV.low.gc','description':'Bottom Decile in GC Content'}
+
+        self.add_section(
+            name = 'Sequencing Depth',
```
Thanks for the comment and showing me this option. In fact the `mu` column is the depth. `cv` column shows the variation. Sequencing depth variation characterize the **sequencing uniformity** and is one of the most important QC metrics in my point of view, particularly for single cell/ low input libraries. Many of the single-cell libraries we have are strongly enriched in coming from specific territories of the genome, like, the region low or high in GC content or peri-centromeric satellites. This has been captured quite well using the `cv` column. I hope that justifies keeping this section?

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+                {'id':'biscuit_isize', 'title': 'Insert Size Distribution',
+                 'ymin': 0, 'yLabelFormat': '{value}%', 
+                 'tt_label': '<strong>Q{point.x}:</strong> {point.y:.2f}% of reads', 
+                 'ylab': '% Mapped Reads', 'xlab': 'Insert Size'}))
```
This is added. Great suggestion!

zwdzwd commented on multiqc/modules/biscuit/biscuit.py:
```diff
@@ -0,0 +1,830 @@
+            name = 'Cytosine Retention',
+            anchor = 'biscuit-retention',
+            description = "This plot shows cytosine retention rate. `r.` stands for read-averaging and `b.` stands for 'base-averaging.",
+            helptext = "**Cytosine retention rate** is 1.0 - **cytosine conversion rate**. Assuming full (complete but not over) bisulfite conversion, **cytosine retention rate** is the average cytosine modification (including 5mC, 5hmC etc) rate.",
```
The format is changed accordingly. Great suggestion!

zwdzwd commented:
Thanks for the very careful code review @ewels!  They are very helpful. I have responded to all the comments and addressed most suggested changes. Let me know of further issues.
